<a href="https://colab.research.google.com/github/KumundzhievMaxim/AppliedDeepLearning/blob/main/EmotionRecognition/emotion_recognition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Downloading the Data {When UNI Link unreacheable}

In [ ]:
# Audio Song 1-24
!wget -O Audio_Song_Actors_01-24.zip https://zenodo.org/record/1188976/files/Audio_Song_Actors_01-24.zip?download=1 

--2020-12-15 07:30:35--  https://zenodo.org/record/1188976/files/Audio_Song_Actors_01-24.zip?download=1
Resolving zenodo.org (zenodo.org)... 137.138.76.77
Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 225505317 (215M) [application/octet-stream]
Saving to: ‘Audio_Song_Actors_01-24.zip’

Audio_Song_Actors_0 100%[===================>] 215.06M  7.43MB/s    in 31s     

2020-12-15 07:31:07 (7.04 MB/s) - ‘Audio_Song_Actors_01-24.zip’ saved [225505317/225505317]



In [ ]:
# Audio Speech
!wget -O Audio_Speech_Actors_01-24.zip https://zenodo.org/record/1188976/files/Audio_Speech_Actors_01-24.zip?download=1

--2020-12-15 07:31:57--  https://zenodo.org/record/1188976/files/Audio_Speech_Actors_01-24.zip?download=1
Resolving zenodo.org (zenodo.org)... 137.138.76.77
Connecting to zenodo.org (zenodo.org)|137.138.76.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 208468073 (199M) [application/octet-stream]
Saving to: ‘Audio_Speech_Actors_01-24.zip’

Audio_Speech_Actors 100%[===================>] 198.81M  8.33MB/s    in 27s     

2020-12-15 07:32:25 (7.41 MB/s) - ‘Audio_Speech_Actors_01-24.zip’ saved [208468073/208468073]



In [ ]:
# Unzip Audio Song 1-24 (./song)
!mkdir ./song && unzip /content/Audio_Song_Actors_01-24.zip -d ./song && rm -r /content/Audio_Song_Actors_01-24.zip

mkdir: cannot create directory ‘./song’: File exists


In [ ]:
# Unzip Audio Speech 1-24 (./speech)
!mkdir ./speech && unzip -q /content/Audio_Speech_Actors_01-24.zip -d ./speech  && rm -r /content/Audio_Speech_Actors_01-24.zip

In [ ]:
# Downalod Video Song Actor 1-9
import os

for video_index in range(1, 10):
  down_cmd = f'wget -O Video_Song_Actor_0{video_index}.zip https://zenodo.org/record/1188976/files/Video_Song_Actor_0{video_index}.zip?download=1'
  dir_cmd = f'mkdir ./video/Actor_0{video_index}'
  unzip_cmd = f'unzip -q Video_Song_Actor_0{video_index}.zip -d ./video/Actor_0{video_index}/'
  rm_cmd = f'rm -r Video_Song_Actor_0{video_index}.zip'

  print(f'Downloading Video_{video_index} zip')
  os.system(down_cmd)
  
  print('Created folder for particular Video zip')
  os.system(dir_cmd)
  
  print('Unzipping particular Video')
  os.system(unzip_cmd)

  print('Deleting particular Video zip')
  os.system(rm_cmd)


Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip
Created folder for particular Video zip
Unzipping particular Video
Deleting particular Video zip


# ENTRY POINT

## Data Description
The RAVDESS contains **7356** files. 
Each file was rated **10 times** on emotional validity, intensity, and genuineness

[Video example](https://www.youtube.com/watch?v=Y7OQoNEu3dY)

The database contains 24 professional actors (12 female, 12 male), vocalizing two lexically-matched statements in a neutral North American accent. Speech includes calm, happy, sad, angry, fearful, surprise, and disgust expressions.


All conditions are available in three modality formats: Audio-only (16bit, 48kHz .wav), Audio-Video (720p H.264, AAC 48kHz, .mp4), and Video-only (no sound).  <b>Note, there are no song files for Actor_18.</b>


### Names convetions and details
#### Audio-only files {Audio_Speech_Actors}
Audio-only files of all actors (01-24) are available as two separate zip files (~200 MB each):

Speech file (Audio_Speech_Actors_01-24.zip, 215 MB) contains 1440 files: 60 trials per actor x 24 actors = 1440. 

#### Audio-Visual and Video-only files files {Video_Speech_Actor}
Speech files (Video_Speech_Actor_01.zip to Video_Speech_Actor_24.zip) collectively contains 2880 files: 60 trials per actor x 2 modalities (AV, VO) x 24 actors = 2880.


### Download Data when UNI Link reachable

In [ ]:
!wget -O ravdess.zip http://nipg1.inf.elte.hu:8765/ravdess.zip 

--2020-12-15 14:52:48--  http://nipg1.inf.elte.hu:8765/ravdess.zip
Resolving nipg1.inf.elte.hu (nipg1.inf.elte.hu)... 157.181.160.161
Connecting to nipg1.inf.elte.hu (nipg1.inf.elte.hu)|157.181.160.161|:8765... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13565818088 (13G) [application/zip]
Saving to: ‘ravdess.zip’

ravdess.zip         100%[===================>]  12.63G  16.2MB/s    in 11m 44s 

2020-12-15 15:04:33 (18.4 MB/s) - ‘ravdess.zip’ saved [13565818088/13565818088]



In [3]:
!unzip -q ravdess.zip 

## Download and Setup **jumcutter**
[Original Documentation](https://github.com/carykh/jumpcutter) 

In [ ]:
!wget -O jumcutter.py https://raw.githubusercontent.com/carykh/jumpcutter/master/jumpcutter.py
!wget -O requirements.txt https://raw.githubusercontent.com/carykh/jumpcutter/master/requirements.txt
!pip install -r requirements.txt

--2020-12-15 07:59:57--  https://raw.githubusercontent.com/carykh/jumpcutter/master/jumpcutter.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 8147 (8.0K) [text/plain]
Saving to: ‘jumcutter.py’

jumcutter.py        100%[===================>]   7.96K  --.-KB/s    in 0s      

2020-12-15 07:59:57 (77.1 MB/s) - ‘jumcutter.py’ saved [8147/8147]

--2020-12-15 07:59:57--  https://raw.githubusercontent.com/carykh/jumpcutter/master/requirements.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 35 [text/plain]
Saving to: ‘requirements.txt’

req

In [ ]:
import os
from tqdm import tqdm 
from typing import List
from pathlib import Path

ROOT_DIR = './ravdess'

video_zips = sorted([zip_name for zip_name in os.listdir(ROOT_DIR) if zip_name.startswith('Video')])  
audio_zips = sorted([zip_name for zip_name in os.listdir(ROOT_DIR) if zip_name.startswith('Audio')])  

def unzip_video_speech(video_zip_paths: List[str]):
  for video_index, video_zip in enumerate(video_zip_paths):  
    source = f'{ROOT_DIR}/{video_zip}'
    target = f'{ROOT_DIR}/videos'
    
    print(f'Unzip {source} TO {target}')

    unzip_cmd = f'unzip -q {source} -d {target} &' # umpersant relies for continue of the execution, otherwise STOPS. 
    os.system(unzip_cmd)
  return True

def delete_zips(paths: List[str]):
  """Permanently delete zips.
  """
  assert [False for path in paths if not path.endswith('*.zip')], 'paths have NOT .zip extensioned files.' 
  
  for path in paths:
    rm_cmd = f'rm -r {path} &' # umpersant relies for continue of the execution, otherwise STOPS. 
    os.system(rm_cmd)
    print(f'Permanently deleted {path}')
  return True


def transform_videos():
  """
  Notes:
    Resulting Folder Structure for Particular Video 
    - Actor_01
      - processed
        processed_0.mp4
        ...
        processed_n.mp4
      01-01-01-01-01-01.mp4
      ...
  Returns:
    proceesed_videos_paths (List[str]): list of full pathes for processed videos
  """

  proceesed_videos_paths = []

  root_videos_path = Path(ROOT_DIR, 'videos') # stored all retrieved video_speech
  video_sources = os.listdir(root_videos_path) # [Actor_01, ..., Actor_24]
  
  # @TODO delete --> reduce number of proccessing folders == 2
  video_sources = video_sources[:2]

  for _, folder in enumerate(tqdm(video_sources)):
    _current = Path(root_videos_path / folder) # [./ravdess/videos/Actor_<>/]
    _current_videos = os.listdir(_current)
    
    # create resulting folder for particular ACTOR folder
    _output_path = f'{_current}/processed/' # [./ravdess/videos/Actor_<>/processed/]
    mkdir_cmd = f'mkdir {_output_path} &'
    os.system(mkdir_cmd)

    for video_index, video in enumerate(tqdm(_current_videos)):
      print(f'Processing {video}, {video_index+1}/{len(_current_videos)}')

      output_file = f'{_output_path}/processed_{video_index}.mp4' 

      # get rid of silience in the beggining and end of the video
      video_silence_cmd = f'python jumcutter.py --input_file {_current}/{video} --output_file {output_file} --sounded_speed 1 --silent_speed 999999 --frame_margin 2 --frame_rate 30'
      os.system(video_silence_cmd)
      
      proceesed_videos_paths.append(output_file)

  return proceesed_videos_paths

FileNotFoundError: ignored

In [ ]:
# -- main
unzip_video_speech(video_zips)
# @TODO add missed step for unzip audio_speech 
processed_videos = transform_videos()

Audio_Speech_Actors_01-24.zip  Video_Speech_Actor_13.zip
Video_Speech_Actor_01.zip      Video_Speech_Actor_14.zip
Video_Speech_Actor_02.zip      Video_Speech_Actor_15.zip
Video_Speech_Actor_03.zip      Video_Speech_Actor_16.zip
Video_Speech_Actor_04.zip      Video_Speech_Actor_17.zip
Video_Speech_Actor_05.zip      Video_Speech_Actor_18.zip
Video_Speech_Actor_06.zip      Video_Speech_Actor_19.zip
Video_Speech_Actor_07.zip      Video_Speech_Actor_20.zip
Video_Speech_Actor_08.zip      Video_Speech_Actor_21.zip
Video_Speech_Actor_09.zip      Video_Speech_Actor_22.zip
Video_Speech_Actor_10.zip      Video_Speech_Actor_23.zip
Video_Speech_Actor_11.zip      Video_Speech_Actor_24.zip
Video_Speech_Actor_12.zip
